core/src/mast/multi_forest_node_iterator.rs

```diff
@@ -0,0 +1,508 @@
+    visited_nodes: Vec<Vec<bool>>,
+    /// This stack always contains the discovered nodes.
+    /// The stack might contain a node twice. However we only ever return a node once, which is
+    /// checked in the `next` function.
+    discovered_nodes: Vec<MultiMastForestIteratorItem>,
+}
+
+impl<'forest> MultiMastForestNodeIter<'forest> {
+    /// Builds a map of MAST roots to non-external nodes in any of the given forests to initialize
+    /// the iterator. This enables an efficient check whether for any encountered External node
+    /// referencing digest `foo` a node with digest `foo` already exists in any forest.
+    pub(crate) fn new(mast_forests: Vec<&'forest MastForest>) -> Self {
+        let visited_nodes = mast_forests
+            .iter()
+            .map(|forest| vec![false; forest.num_nodes() as usize])
+            .collect();
+
+        let mut non_external_nodes = BTreeMap::new();
+
+        for (forest_idx, forest) in mast_forests.iter().enumerate() {
+            for (node_idx, node) in forest.nodes().iter().enumerate() {
+                // SAFETY: The passed id comes from the iterator over the nodes, so we never exceed
+                // the forest's number of nodes.
+                let node_id = MastNodeId::new_unsafe(node_idx as u32);
+                if !node.is_external() {
+                    non_external_nodes.insert(node.digest(), (forest_idx, node_id));
+                }
+            }
+        }
+
+        Self {
+            mast_forests,
+            current_forest_idx: 0,
+            current_procedure_root_idx: 0,
+            non_external_nodes,
+            visited_nodes,
+            discovered_nodes: Vec::new(),
+        }
+    }
+
+    /// Pushes the given node, uniquely identified by the forest and node index onto the stack
+    /// even if the node was already discovered before.
+    ///
+    /// It's the callers responsibility to only pass valid indices.
+    fn push_node(&mut self, forest_idx: usize, node_id: MastNodeId) {
+        self.discovered_nodes
+            .push(MultiMastForestIteratorItem::Node { forest_idx, node_id });
+    }
+
+    /// Discovers a tree starting at the given forest index and node id.
+    ///
+    /// It's the callers responsibility to only pass valid indices.
+    fn discover_tree(
+        &mut self,
+        forest_idx: ForestIndex,
+        node_id: MastNodeId,
+    ) -> Result<(), MastForestError> {
+        let current_node =
+            &self.mast_forests[forest_idx].nodes.get(node_id.as_usize()).ok_or_else(|| {
+                MastForestError::NodeIdOverflow(
+                    node_id,
+                    self.mast_forests[forest_idx].num_nodes() as usize,
+                )
+            })?;
+
+        // Note that the order in which we add or discover nodes is the reverse of postorder, since
+        // we're pushing them onto a stack, which reverses the order itself. Hence, reversing twice
```

Comment:
Reversing just _once_, right? You visit the tree in preorder pushing nodes on the visit stack, and by popping the nodes off the stack, you get them in postorder (and more precisely, you are guaranteed to pop child nodes before their parents).

Note, the only part of MAST which is a tree are leaf procedures (or any procedure if you do not follow calls to their callees). A `MastForest` is actually a DAG, and depending on how you are navigating it, it _can_ have cycles (i.e. if you are tracing procedure references, since `procref` admits recursion). All that to say, we have to be a bit careful about assuming certain tree/graph properties, depending on what we're doing). As I've noted, we can't assume that when we see a child, that it is the _first_ time we've seen it (something we could assume of a tree), because it may have been visited by another root (we're actually working with a graph/forest).

Not saying anything here is incorrect, but mostly to try and ensure the commentary isn't confusing for future readers.

Reply:
Thanks for the explanation!

I can totally see why the reverse comment was confusing. It's no longer necessary with the move to `VecDeque` so I'll resolve this.